### DISCUSSION TITLE
fenced_divs with preformatted text (source code)

### DISCUSSION THREAD

**Discussion by lifeisfoo** [jgm/pandoc#10489] [Q&A] [answered]
I'm using `fenced_divs` to wrap code blocks in order to match a custom style in the output file (**markdown->docx**).

```
::: {custom-style="MyCodeStyle"}
function hello() {  
    console.log("world");  
}
:::
```

Pandoc reads that block as a paragraph, so spaces are removed.

```
    {
      "t": "Div",
      "c": [
        ["", [], [["custom-style", "MyCodeStyle"]]],
        [
          {
            "t": "Para",
            "c": [
              { "t": "Str", "c": "function" },
              { "t": "Space" },
              { "t": "Str", "c": "hello()" },
              { "t": "Space" },
              { "t": "Str", "c": "{" },
              { "t": "LineBreak" },
              { "t": "Str", "c": "console.log(" },
              {
                "t": "Quoted",
                "c": [{ "t": "DoubleQuote" }, [{ "t": "Str", "c": "world" }]]
              },
              { "t": "Str", "c": ");" },
              { "t": "LineBreak" },
              { "t": "Str", "c": "}" }
            ]
          }
        ]
      ]
    },
```

Is there a way to force pandoc `fenced_divs` to behave like a code block producing an output like this?

```
    {
      "t": "CodeBlock",
      "c": [
        ["", [], []],
        "function hello() {  \n    console.log(\"world\");  \n}  "
      ]
    }
```

I saw [this solution](https://github.com/jgm/pandoc/issues/10077#issuecomment-2282804741) but I would like to avoid converting all spaces to non breaking ones.

**Comment by jgm** [comment]
Why not put it in a code block and attach the custom style to that?
````
``` {custom-style="MyCodeStyle"}
function hello() {  
    console.log("world");  
}
```
````

**Reply by lifeisfoo** [reply]
Thank you @jgm.
I was not aware of this. Reading [the doc](https://pandoc.org/MANUAL.html#output) this is something not obvious.

The AST now contains the custom style but it is not present in the final docx (the style used is the default "Source Code").

```
    {
      "t": "CodeBlock",
      "c": [
        [
          "",
          [],
          [
            ["custom-style", "MyCodeStyle"]
          ]
        ],
        "function hello() {  \n    console.log(\"world\");  \n}  "
      ]
    },
```

![immagine](https://github.com/user-attachments/assets/fc2b04cb-612b-4929-9723-2c5980f0bfdd)

**Reply by jgm** [reply]
OK, maybe custom styles only work with certain elements; I had forgotten the details.
Unfortunately you can't apply multiple styles in Word.

**Comment by jgm** [comment]
You could try something like
````
::: {custom-style="MyCodeStyle"}
`function hello() {`\
`    console.log("world");`\
`}`\
:::
````

**Reply by lifeisfoo** [reply]
It seems that spaces are being trimmed even with this.

```
    {
      "t": "Div",
      "c": [
        ["", [], [["custom-style", "MyCodeStyle"]]],
        [
          {
            "t": "Para",
            "c": [
              { "t": "Code", "c": [["", [], []], "function hello() {"] },
              { "t": "LineBreak" },
              { "t": "Code", "c": [["", [], []], "console.log(\"world\");"] },
              { "t": "LineBreak" },
              { "t": "Code", "c": [["", [], []], "}"] },
              { "t": "LineBreak" }
            ]
          }
        ]
      ]
    },
```

If there is no way of achieving this, I will end up using the plain code block as you suggested in the first reply adapting the "Source Code" style in the reference doc (not my preferred choice since I depend on third party styles/template).

In some cases, forcing the use of a single style when dealing with preformatted text could be limiting.
Having the support for custom-style in Code Blocks will add coherence in pandoc's features (since they are very similar to fenced divs).

**Reply by jgm** [reply]
Yeah, I forgot that the default `markdown` will trim spaces at the beginning and end of code spans.  Commonmark won't do that, though.  With `commonmark_x`:
```
[ Div
    ( "" , [] , [ ( "custom-style" , "MyCodeStyle" ) ] )
    [ Para
        [ Code ( "" , [] , [] ) "function hello() {"
        , LineBreak
        , Code ( "" , [] , [] ) "    console.log(\"world\");"
        , LineBreak
        , Code ( "" , [] , [] ) "}"
        , Str "\\"
        ]
    ]
]
```

Note: the "single style" constraint is not from pandoc, it's from Word.  Word's styles don't cascade like in CSS.  We could support custom-style in code blocks, but that would mean not using the SourceCode style, so your custom style would have to reproduce what it does.

**Reply by jgm** [reply, marked as answer]
From the manual:

> If you define a `div` or `span` with the attribute `custom-style`,
pandoc will apply your specified style to the contained elements (with
the exception of elements whose function depends on a style, like
headings, code blocks, block quotes, or links).

In Word, everything (except a table) is a paragraph; headings and code blocks and block quotes just differ in style.  That's why we don't apply custom styles to code blocks.
